src/main/java/io/github/moulberry/notenoughupdates/profileviewer/SacksPage.kt

```diff
@@ -0,0 +1,505 @@
+
+    private val sackPattern = "^RUNE_(?<name>\\w+)_(?<tier>\\d)\$".toPattern()
+
+    override fun drawPage(mouseX: Int, mouseY: Int, partialTicks: Float) {
+        guiLeft = GuiProfileViewer.getGuiLeft()
+        guiTop = GuiProfileViewer.getGuiTop()
+
+        MC.textureManager.bindTexture(pv_sacks)
+        Utils.drawTexturedRect(
+            guiLeft.toFloat(),
+            guiTop.toFloat(),
+            instance.sizeX.toFloat(),
+            instance.sizeY.toFloat(),
+            GL11.GL_NEAREST
+        )
+
+        val newProfile = selectedProfile
+        if (newProfile == null) {
+            Utils.drawStringCentered("§cMissing Profile Data", guiLeft + 250, guiTop + 101, true, 0)
+            return
+        }
+
+        if (sacksJson == null) {
+            Utils.drawStringCentered("§cMissing Repo Data", guiLeft + 250, guiTop + 101, true, 0)
+            return
+        }
+
+        if (newProfile != currentProfile) {
+            getData()
+            currentProfile = selectedProfile
+        }
+
+        val currentSackData = sackContents[currentSack] ?: run {
+            Utils.drawStringCentered("§cApi Info Missing", guiLeft + 250, guiTop + 101, true, 0)
+            return
+        }
+        // after this point everything in the constants json exists and does not need to be checked again except "item"
+
+        val name = if (currentSack == "All") "§2All Sacks" else "§2$currentSack Sack"
+        Utils.renderShadowedString(name, (guiLeft + 78).toFloat(), (guiTop + 74).toFloat(), 105)
+
+        Utils.renderAlignedString(
+            "§6Value",
+            "§f${StringUtils.formatNumber(currentSackData.sackValue.toLong())}",
+            (guiLeft + 27).toFloat(),
+            (guiTop + 91).toFloat(),
+            102
+        )
+
+        Utils.renderAlignedString(
+            "§2Items",
+            "§f${StringUtils.formatNumber(currentSackData.itemCount)}",
+            (guiLeft + 27).toFloat(),
+            (guiTop + 108).toFloat(),
+            102
+        )
+
+        GlStateManager.enableDepth()
+
+        var xIndex = 0
+        var yIndex = 0
+
+        val sackTypes = sacksJson.getAsJsonObject("sacks")
+
+        val startIndex = page * pageSize
+        val endIndex = (page + 1) * pageSize
+        var itemCount = -1
+        if (currentSack == "All") {
+            for ((sackName, sackData) in sackTypes.entrySet()) {
+                itemCount++
+                if (itemCount < startIndex || itemCount >= endIndex) continue
+
+                if (yIndex == rows) continue
+                val data = sackData.asJsonObject
+
+                if (!data.has("item") || !data.get("item").isJsonPrimitive || !data.get("item").asJsonPrimitive.isString) continue
+                val sackItemName = data.get("item").asString
+                val itemStack = manager.createItem(sackItemName)
+
+                val x = guiLeft + 168 + xIndex * 37
+                val y = guiTop + 20 + yIndex * 41
```

Comment:
Can we move this positioning calculation into a method? Having magic numbers that need to be updated in multiple places just begs for bad behaviour down the line.